content/blog/2022-05-19-hcl-mlir/index.md

```diff
@@ -0,0 +1,514 @@
+          %4 = arith.addi %2, %3 : i32
+          %5 = affine.load %1[1, %arg2 + 1] : memref<3x10xi32>
+          %6 = arith.addi %4, %5 : i32
+          %7 = affine.load %1[1, %arg2 + 2] : memref<3x10xi32>
+          %8 = arith.addi %6, %7 : i32
+          %9 = affine.load %1[2, %arg2 + 1] : memref<3x10xi32>
+          %10 = arith.addi %8, %9 : i32
+          affine.store %10, %0[%arg1 - 2, %arg2] : memref<8x8xi32>
+        } {loop_name = "j"}
+      }
+    } {loop_name = "i", stage_name = "B"}
+    return %0 : memref<8x8xi32>
+  }
+}
+```
+
+We also add additional optimizations to make generated code more efficient. One of them is to merge loops with the same bound. The following example shows the five-point stencil code with line buffer and window buffer, and we can see that the shifting part of the line buffer is merged with the inner computation part (only generate one `j` loop in this example), so that the loop overhead is reduced.
+
+```cpp
+#set = affine_set<(d0) : (d0 - 2 >= 0)>
+module {
+  func @top(%arg0: memref<10x10xi32>) -> memref<8x8xi32> {
+    %0 = memref.alloc() {name = "B"} : memref<8x8xi32>
+    %1 = memref.alloc() : memref<3x10xi32>
+    %2 = memref.alloc() : memref<3x3xi32>
+    affine.for %arg1 = 0 to 10 {
+      affine.for %arg2 = 0 to 10 {
+        %3 = affine.load %1[1, %arg2] : memref<3x10xi32>
+        affine.store %3, %1[0, %arg2] : memref<3x10xi32>
+        %4 = affine.load %1[2, %arg2] : memref<3x10xi32>
+        affine.store %4, %1[1, %arg2] : memref<3x10xi32>
+        %5 = affine.load %arg0[%arg1, %arg2] : memref<10x10xi32>
+        affine.store %5, %1[2, %arg2] : memref<3x10xi32>
+        affine.if #set(%arg1) {
+          affine.for %arg3 = 0 to 3 {
+            %6 = affine.load %2[%arg3, 1] : memref<3x3xi32>
+            affine.store %6, %2[%arg3, 0] : memref<3x3xi32>
+            %7 = affine.load %2[%arg3, 2] : memref<3x3xi32>
+            affine.store %7, %2[%arg3, 1] : memref<3x3xi32>
+            %8 = affine.load %1[%arg3, %arg2] : memref<3x10xi32>
+            affine.store %8, %2[%arg3, 2] : memref<3x3xi32>
+          } {spatial}
+          affine.if #set(%arg2) {
+            %6 = affine.load %2[1, 0] : memref<3x3xi32>
+            %7 = affine.load %2[0, 1] : memref<3x3xi32>
+            %8 = arith.addi %6, %7 : i32
+            %9 = affine.load %2[1, 1] : memref<3x3xi32>
+            %10 = arith.addi %8, %9 : i32
+            %11 = affine.load %2[2, 1] : memref<3x3xi32>
+            %12 = arith.addi %10, %11 : i32
+            %13 = affine.load %2[1, 2] : memref<3x3xi32>
+            %14 = arith.addi %12, %13 : i32
+            affine.store %14, %0[%arg1 - 2, %arg2 - 2] : memref<8x8xi32>
+          }
+        }
+      } {loop_name = "j"}
+    } {loop_name = "i", stage_name = "B"}
+    return %0 : memref<8x8xi32>
+  }
+}
+```
+
+The above steps can be extended to high-dimensional stencils, and we will give more experimental results in the following sections.
```

Comment:
Overall, this section makes me wonder whether—in an ideal world—you might have preferred a different program representation (i.e., not the `affine` dialect)? I'm not entirely clear on the distinction, but there seem to be some challenges here that are fundamental and some that are a consequence of the program representation.

Reply:
Right, I think the reason why affine dialect decouples the affine map from operands is that it is easier to do affine transformation. Some facilities for polyhedral compilation should exist in MLIR, but unfortunately, we have not found a good one to incorporate, so we can only write the analysis part ourselves, which is awkward.

Reply:
Right, I think the reason why affine dialect decouples the affine map from operands is that it is easier to do affine transformation. Some facilities for polyhedral compilation should exist in MLIR, but unfortunately, we have not found a good one to incorporate, so we can only write the analysis part ourselves, which is awkward.

Reply:
TVM can represent straight one-line code in their IR like `B[i] = A[i] + A[i+1] + A[i+2]`. This may simplify some analysis, but complicate other transformations. Different representations seem to have tradeoff here.